File: app/src/main/java/za/co/woolworths/financial/services/android/ui/fragments/wreward/WRewardsVouchersFragment.java
package za.co.woolworths.financial.services.android.ui.fragments.wreward;

import android.app.Activity;
import android.content.Context;
import android.content.Intent;
import android.os.Bundle;
import android.support.annotation.Nullable;
import android.support.v4.app.Fragment;
import android.support.v7.widget.LinearLayoutManager;
import android.support.v7.widget.RecyclerView;
import android.view.LayoutInflater;
import android.view.View;
import android.view.ViewGroup;
import android.widget.ImageView;
import android.widget.RelativeLayout;

import com.awfs.coordination.R;
import com.google.gson.Gson;

import io.reactivex.android.schedulers.AndroidSchedulers;
import io.reactivex.disposables.CompositeDisposable;
import io.reactivex.functions.Consumer;
import io.reactivex.schedulers.Schedulers;
import za.co.woolworths.financial.services.android.contracts.FirebaseManagerAnalyticsProperties;
import za.co.woolworths.financial.services.android.models.WoolworthsApplication;
import za.co.woolworths.financial.services.android.models.dto.VoucherResponse;
import za.co.woolworths.financial.services.android.ui.activities.WRewardsVoucherDetailsActivity;
import za.co.woolworths.financial.services.android.ui.adapters.WRewardsVoucherListAdapter;
import za.co.woolworths.financial.services.android.ui.views.ScrollingLinearLayoutManager;
import za.co.woolworths.financial.services.android.ui.views.WTextView;
import za.co.woolworths.financial.services.android.util.AuthenticateUtils;
import za.co.woolworths.financial.services.android.util.ErrorHandlerView;
import za.co.woolworths.financial.services.android.util.RecycleViewClickListner;
import za.co.woolworths.financial.services.android.util.Utils;

/**
 * Created by W7099877 on 05/01/2017.
 */

public class WRewardsVouchersFragment extends Fragment {
	private ScrollingLinearLayoutManager mLayoutManager;
	private WRewardsVoucherListAdapter mAdapter;
	private RecyclerView recyclerView;
	public VoucherResponse voucherResponse;
	private ErrorHandlerView mErrorHandlerView;
	public int selectedVoucherPosition;
	public static final int LOCK_REQUEST_CODE_WREWARDS = 111;
	private CompositeDisposable mDisposables = new CompositeDisposable();
	private Activity mActivity;
	private boolean isAuthenticated;

	@Nullable
	@Override
	public View onCreateView(LayoutInflater inflater, @Nullable ViewGroup container, @Nullable Bundle savedInstanceState) {
		View view = inflater.inflate(R.layout.wrewards_vouchers_fragment, container, false);
		Utils.triggerFireBaseEvents(FirebaseManagerAnalyticsProperties.WREWARDSVOUCHERS);
		Bundle bundle = getArguments();
		voucherResponse = new Gson().fromJson(bundle.getString("WREWARDS"), VoucherResponse.class);
		recyclerView = (RecyclerView) view.findViewById(R.id.recycler_view);

		mErrorHandlerView = new ErrorHandlerView(getActivity(),
				(RelativeLayout) view.findViewById(R.id.relEmptyStateHandler),
				(ImageView) view.findViewById(R.id.imgEmpyStateIcon),
				(WTextView) view.findViewById(R.id.txtEmptyStateTitle),
				(WTextView) view.findViewById(R.id.txtEmptyStateDesc));

		mLayoutManager = new ScrollingLinearLayoutManager(
				getActivity(),
				LinearLayoutManager.VERTICAL,
				false, 1500
		);

		recyclerView.setLayoutManager(mLayoutManager);

		if (voucherResponse.voucherCollection.vouchers == null || voucherResponse.voucherCollection.vouchers.size() == 0) {
			displayNoVouchersView();
		} else {
			displayVouchers(voucherResponse);
		}

		mDisposables.add(WoolworthsApplication.getInstance()
				.bus()
				.toObservable()
				.subscribeOn(Schedulers.io())
				.observeOn(AndroidSchedulers.mainThread())
				.subscribe(new Consumer<Object>() {
					@Override
					public void accept(Object object) throws Exception {
						if (object != null) {
							if (object instanceof WRewardsVouchersFragment) {
								if (!isAuthenticated) {
									AuthenticateUtils.getInstance(getActivity()).enableBiometricForCurrentSession(false);
									startVoucherDetailsActivity();
									isAuthenticated = true;
								}
							}
						}
					}
				}));
		return view;
	}

	public void displayNoVouchersView() {
		mErrorHandlerView.showEmptyState(0);
		mErrorHandlerView.hideIcon();
		mErrorHandlerView.hideTitle();
		mErrorHandlerView.textDescription(getActivity().getResources().getString(R.string.no_vouchers));
		recyclerView.setVisibility(View.GONE);
	}


	public void displayVouchers(final VoucherResponse vResponse) {
		mErrorHandlerView.hideEmpyState();
		recyclerView.setVisibility(View.VISIBLE);
		mAdapter = new WRewardsVoucherListAdapter(getActivity(), vResponse.voucherCollection.vouchers);
		recyclerView.setAdapter(mAdapter);
		recyclerView.addOnItemTouchListener(new RecycleViewClickListner(getActivity(), recyclerView, new RecycleViewClickListner.ClickListener() {
			@Override
			public void onClick(View view, int position) {
				selectedVoucherPosition = position;
				if (AuthenticateUtils.getInstance(getActivity()).isBiometricAuthenticationRequired()) {
					try {
						AuthenticateUtils.getInstance(getActivity()).startAuthenticateApp(LOCK_REQUEST_CODE_WREWARDS);
					} catch (Exception e) {
						e.printStackTrace();
					}
				} else {
					startVoucherDetailsActivity();
				}

			}

			@Override
			public void onLongClick(View view, int position) {
			}
		}));
	}

	public void startVoucherDetailsActivity() {
		Intent intent = new Intent(mActivity, WRewardsVoucherDetailsActivity.class);
		intent.putExtra("VOUCHERS", Utils.objectToJson(voucherResponse.voucherCollection));
		intent.putExtra("POSITION", selectedVoucherPosition);
		mActivity.startActivity(intent);
	}

	@Override
	public void onAttach(Context context) {
		super.onAttach(context);
		if (context instanceof Activity) {
			mActivity = (Activity) context;
		}
	}

	public void scrollToTop() {
		recyclerView.smoothScrollToPosition(0);
	}
}

File: app/src/main/java/za/co/woolworths/financial/services/android/ui/fragments/wreward/WRewardsVouchersFragment.kt
package za.co.woolworths.financial.services.android.ui.fragments.wreward

import android.content.Intent
import android.os.Bundle
import android.view.View
import android.widget.RelativeLayout
import androidx.fragment.app.Fragment
import androidx.recyclerview.widget.LinearLayoutManager
import androidx.recyclerview.widget.RecyclerView
import com.awfs.coordination.R
import com.google.gson.Gson
import dagger.hilt.android.AndroidEntryPoint
import za.co.woolworths.financial.services.android.contracts.FirebaseManagerAnalyticsProperties
import za.co.woolworths.financial.services.android.models.dto.VoucherResponse
import za.co.woolworths.financial.services.android.ui.activities.WRewardsVoucherDetailsActivity
import za.co.woolworths.financial.services.android.ui.activities.dashboard.BottomNavigationActivity
import za.co.woolworths.financial.services.android.ui.adapters.WRewardsVoucherListAdapter
import za.co.woolworths.financial.services.android.ui.views.ScrollingLinearLayoutManager
import za.co.woolworths.financial.services.android.ui.wfs.common.biometric.AuthenticateUtils
import za.co.woolworths.financial.services.android.ui.wfs.common.biometric.BiometricCallback
import za.co.woolworths.financial.services.android.ui.wfs.common.biometric.WfsBiometricManager
import za.co.woolworths.financial.services.android.util.ErrorHandlerView
import za.co.woolworths.financial.services.android.util.RecycleViewClickListner
import za.co.woolworths.financial.services.android.util.Utils
import javax.inject.Inject

@AndroidEntryPoint
class WRewardsVouchersFragment : Fragment(R.layout.wrewards_vouchers_fragment) {

    private var vouchersRecyclerView: RecyclerView? = null
    private var voucherResponse: VoucherResponse? = null
    private var mErrorHandlerView: ErrorHandlerView? = null
    var selectedVoucherPosition = 0
    private var relEmptyStateHandler: RelativeLayout? = null
    private var isBiometricAuthenticated : Boolean = false
    @Inject lateinit var biometricManager : WfsBiometricManager

    override fun onViewCreated(view: View, savedInstanceState: Bundle?) {
        super.onViewCreated(view, savedInstanceState)
        voucherResponse = Gson().fromJson(arguments?.getString(WREWARDS), VoucherResponse::class.java)
        initView(view)
        initVouchers()
        setUpBiometricAuthentication()
        uniqueIdsForRewardVoucherAutomation()
    }

    private fun setUpBiometricAuthentication() {
        val bottomNavigationActivity =  (requireActivity() as? BottomNavigationActivity)
        biometricManager.setupBiometricInWRewardsVouchersOnItemTap(fragment = this,
            bottomNavigation = bottomNavigationActivity?.bottomNavigationById) { callback ->
            when (callback) {
                BiometricCallback.ErrorUserCanceled -> {
                    bottomNavigationActivity?.bottomNavigationById?.currentItem = BottomNavigationActivity.INDEX_TODAY
                    AuthenticateUtils.enableBiometricForCurrentSession(true)
                }
                BiometricCallback.Succeeded -> {
                      if (!isBiometricAuthenticated) {
                          presentVoucherDetail()
                      }
                }
                else -> Unit
            }
        }
    }

    private fun presentVoucherDetail() {
        AuthenticateUtils.enableBiometricForCurrentSession(false)
        startVoucherDetailsActivity()
        isBiometricAuthenticated = true
    }

    private fun initView(view: View) {
        vouchersRecyclerView = view.findViewById(R.id.vouchersRecyclerView)
        relEmptyStateHandler = view.findViewById(R.id.relEmptyStateHandler)
        mErrorHandlerView = ErrorHandlerView(
            requireActivity(),
            relEmptyStateHandler,
            view.findViewById(R.id.imgEmpyStateIcon),
            view.findViewById(R.id.txtEmptyStateTitle),
            view.findViewById(R.id.txtEmptyStateDesc)
        )
        val mLayoutManager = ScrollingLinearLayoutManager(
            requireActivity(),
            LinearLayoutManager.VERTICAL,
            false, 1500
        )
        vouchersRecyclerView?.layoutManager = mLayoutManager
    }

    private fun initVouchers() {
        val vouchers = voucherResponse?.voucherCollection?.vouchers
        if (vouchers.isNullOrEmpty() || vouchers.size == 0) {
            presentEmptyVoucherView()
        } else {
            presentVouchersListView(this, voucherResponse)
        }
    }

    private fun uniqueIdsForRewardVoucherAutomation() {
        if (requireActivity().resources != null) {
            vouchersRecyclerView?.contentDescription = getString(R.string.vouchersLayout)
            relEmptyStateHandler?.contentDescription = getString(R.string.voucher_empty_state)
        }
    }

    override fun onResume() {
        super.onResume()
        Utils.setScreenName(requireActivity(),
            FirebaseManagerAnalyticsProperties.ScreenNames.WREWARDS_VOUCHERS
        )
    }

    private fun presentEmptyVoucherView() {
        mErrorHandlerView?.apply {
            showEmptyState(0)
            hideIcon()
            hideTitle()
            textDescription(getString(R.string.no_vouchers))
        }
        vouchersRecyclerView?.visibility = View.GONE
    }

    private fun presentVouchersListView(fragment: Fragment, vResponse: VoucherResponse?) {
        mErrorHandlerView?.hideEmpyState()
        val voucherListAdapter = WRewardsVoucherListAdapter()
        voucherListAdapter.setItem(vResponse?.voucherCollection?.vouchers ?: listOf())
        vouchersRecyclerView?.apply {
            visibility = View.VISIBLE
            adapter = voucherListAdapter
        addOnItemTouchListener(
            RecycleViewClickListner(requireActivity(), this, object : RecycleViewClickListner.ClickListener {
                    override fun onClick(view: View, position: Int) {
                        selectedVoucherPosition = position
                        if (biometricManager.isBiometricEnabled(fragment.requireContext()) && !isBiometricAuthenticated) {
                            biometricManager.show()
                        }else {
                            presentVoucherDetail()
                        }
                    }
                    override fun onLongClick(view: View, position: Int) {}
                })
        )
        }

    }

    private fun startVoucherDetailsActivity() {
        val intent = Intent(requireActivity(), WRewardsVoucherDetailsActivity::class.java)
        intent.putExtra(VOUCHERS, Utils.objectToJson(voucherResponse?.voucherCollection))
        intent.putExtra(POSITION, selectedVoucherPosition)
        startActivity(intent)
    }

    fun scrollToTop() {
        vouchersRecyclerView?.smoothScrollToPosition(0)
    }

    companion object {
        const val LOCK_REQUEST_CODE_WREWARDS = 111
        const val VOUCHERS = "VOUCHERS"
        const val POSITION = "POSITION"
        const val WREWARDS = "WREWARDS"
    }
}